pkg/kubelet/kuberuntime/kuberuntime_image.go

```diff
@@ -32,14 +34,25 @@ import (
 
 // PullImage pulls an image from the network to local storage using the supplied
 // secrets if necessary.
-func (m *kubeGenericRuntimeManager) PullImage(ctx context.Context, image kubecontainer.ImageSpec, pullSecrets []v1.Secret, podSandboxConfig *runtimeapi.PodSandboxConfig) (string, error) {
+func (m *kubeGenericRuntimeManager) PullImage(ctx context.Context, image kubecontainer.ImageSpec, pullSecrets []v1.Secret, podSandboxConfig *runtimeapi.PodSandboxConfig, serviceAccountName string) (string, error) {
 	img := image.Image
 	repoToPull, _, _, err := parsers.ParseImageName(img)
 	if err != nil {
 		return "", err
 	}
 
-	keyring, err := credentialprovidersecrets.MakeDockerKeyring(pullSecrets, m.keyring)
+	// construct the dynamic keyring using the providers we have in the kubelet
+	var podName, podNamespace, podUID string
+	if utilfeature.DefaultFeatureGate.Enabled(features.KubeletServiceAccountTokenForCredentialProviders) {
+		if podSandboxConfig != nil && podSandboxConfig.Metadata != nil {
+			podName = podSandboxConfig.Metadata.Name
+			podNamespace = podSandboxConfig.Metadata.Namespace
+			podUID = podSandboxConfig.Metadata.Uid
+		}
+	}
+	externalCredentialProviderKeyring := credentialproviderplugin.NewExternalCredentialProviderDockerKeyring(podName, podNamespace, podUID, serviceAccountName)
+
+	keyring, err := credentialprovidersecrets.MakeDockerKeyring(pullSecrets, credentialprovider.UnionDockerKeyring{m.keyring, externalCredentialProviderKeyring})
```

Comment:
Should the credentials of the node-global keyring have precedence to the SA-specific creds? The ordering of credentials returned from the `Lookup()` method is a bit odd:
1. pod-specific image pull secrets
2. node-global credentials
3. service-account specific creds

I think I'd expect ordering from the most specific (secrets) to the least specific creds (node), that is 1. 3. 2.